src/ciphers/morse_code.rs

```diff
@@ -0,0 +1,206 @@
+        "4" => "....-",   "5" => ".....",   "6" => "-....",
+        "7" => "--...",   "8" => "---..",   "9" => "----.",
+        "0" => "-----",
+
+        "&" => ".-...",   "@" => ".--.-.",  ":" => "---...",
+        "," => "--..--",  "." => ".-.-.-",  "'" => ".----.",
+        "\"" => ".-..-.", "?" => "..--..",  "/" => "-..-.",
+        "=" => "-...-",   "+" => ".-.-.",   "-" => "-....-",
+        "(" => "-.--.",   ")" => "-.--.-",  " " => "/",
+        "!" => "-.-.--",
+    }
+}
+
+fn _morse_to_alphanumeric_dictionary() -> HashMap<&'static str, &'static str> {
+    map! {
+        ".-"   =>  "A",      "-..." => "B",    "-.-." => "C",
+        "-.."  =>  "D",      "."    => "E",       "..-." => "F",
+        "--."  =>  "G",      "...." => "H",    ".." => "I",
+        ".---" =>  "J",     "-.-" => "K",     ".-.." => "L",
+        "--"   =>  "M",       "-." => "N",      "---" => "O",
+        ".--." =>  "P",     "--.-" => "Q",    ".-." => "R",
+        "..."  =>  "S",      "-" => "T",       "..-" => "U",
+        "...-" =>  "V",     ".--" => "W",     "-..-" => "X",
+        "-.--" =>  "Y",     "--.." => "Z",
+
+        ".----" => "1",    "..---" => "2",   "...--" => "3",
+        "....-" => "4",    "....." => "5",   "-...." => "6",
+        "--..." => "7",    "---.." => "8",   "----." => "9",
+        "-----" => "0",
+
+        ".-..." => "&",    ".--.-." => "@",  "---..." => ":",
+        "--..--" => ",",   ".-.-.-" => ".",  ".----." => "'",
+        ".-..-." => "\"",  "..--.." => "?",  "-..-." => "/",
+        "-...-" => "=",   ".-.-." => "+",   "-....-" => "-",
+        "-.--." => "(",   "-.--.-" => ")",  "/" => " ",
+        "-.-.--" => "!",  " " => " ",       "" => ""
+    }
+}
+
+fn _check_part(string: &str) -> bool {
+    for c in string.chars() {
+        match c {
+            '.' | '-' | ' ' => (),
+            _ => return false,
+        }
+    }
+    true
+}
+
+fn _check_all_parts(string: &str) -> bool {
+    string.split('/').all(_check_part)
+}
+
+fn _decode_token(string: &str) -> String {
+    _morse_to_alphanumeric_dictionary()
+        .get(string)
+        .unwrap_or(&_UNKNOWN_MORSE_CHARACTER)
+        .to_string()
+}
+
+fn _decode_part(string: &str) -> String {
+    string
+        .split(' ')
+        .map(_decode_token)
+        .collect::<Vec<String>>()
+        .join("")
+}
+
+/// Convert morse code to ascii.
+///
+/// Given a morse code, return the corresponding message.
+/// If the code is invalid, the undecipherable part of the code is replaced by `_`.
+pub fn decode(string: &str) -> Result<String, io::Error> {
+    if !_check_all_parts(string) {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            "Invalid morse code",
+        ));
+    }
+
+    let mut partitions: Vec<String> = vec![];
+
+    for part in string.split('/') {
+        partitions.push(_decode_part(part));
+    }
+
+    Ok(partitions.join(" "))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
```

Comment:
Maybe by explicitly importing `decode` here we can remove the warning?

Reply:
This doesn't fix it, no

Reply:
Yeah, I think having `#[cfg(test)]` is enough. I also thought of moving `decode` into the `tests` module. Your call.

Reply:
well now it doesn't complain, so :woman_shrugging: no need to change.